[scripts] Introduce packaging scripts to support binary distribution (#124)
mapbox/mapbox-maps-ios: scripts/release/packager/PACKAGE.md

# Packaging MapboxMaps.xcframework

> This folder contains scripts that package `MapboxMaps` as an xcframework.

Reviewer: ```suggestion
This folder contains scripts that package `MapboxMaps` and its dependencies as a zipped archive of XCFrameworks.
```

## Requirements
- [`jq`](https://stedolan.github.io/jq/)
- [`xcodegen`](https://github.com/yonaskolb/XcodeGen) 
- valid `.netrc` token for sdk registry

Descriptions of the files in this directory is as follows:

### `package-mapbox-maps-xcframework.sh`
- This is the master script and is the only one you should need to call.

> - It reads the dependency versions from `versions.json` and begins by downloading previously built xcframeworks & cloning and building any other xcframeworks (currently `Turf`).

Reviewer: ```suggestion
- It reads the dependency versions from `versions.json` and begins by downloading previously built XCFrameworks & cloning and building any other XCFrameworks (currently `Turf`).
```

> - It also creates a just-in-time xcodeproj in order to create a binary xcframework for `MapboxMaps`.

Reviewer: ```suggestion
- It also creates a just-in-time Xcode project in order to create a binary XCFramework for `MapboxMaps`.
```

- NOTE: This script **MUST** be called from this directory 

### `download-dependency-xcframeworks.sh`
- This script downloads `MapboxCoreMaps.xcframework`, `MapboxCommon.xcframework` and `MapboxMobileEvents.xcframework` from SDK Registry
- NOTE: You **MUST** have a valid `.netrc` token

### `create-turf-xcframework.sh`

> - This script clones `turf-swift`, checks out the git tag represented in `versions.json` and then creates a `Turf.xcframework`.

Reviewer: ```suggestion
- This script clones `turf-swift`, checks out the git tag represented in `versions.json` and then creates `Turf.xcframework`.
```

> 

Reviewer: ```suggestion
```

### `create-mapbox-maps-xcframework.sh`
- Helper script to archive and build `MapboxMaps.xcframework`
- Uses `xcodebuild`

### `project.yml`

> - This `yml` is parsed by [`xcodegen`](https://github.com/yonaskolb/XcodeGen) to create a just-in-time xcodeproj that should contain all the sources used in `MapboxMaps`.

Reviewer: ```suggestion
- This `yml` is parsed by [`xcodegen`](https://github.com/yonaskolb/XcodeGen) to create a just-in-time Xcode project that should contain all the sources used in `MapboxMaps`.
```

### `versions.json`

> - This is the "source of truth" that is used to download/build the correct dependency versions

Reviewer: ```suggestion
- This is the "source of truth" that is used to download the correct dependency versions
```

- [`jq`](https://stedolan.github.io/jq/)
- [`xcodegen`](https://github.com/yonaskolb/XcodeGen)
- Valid `.netrc` token for SDK registry

## Scripts

### `package-mapbox-maps.sh`

- Usage:
  ```

>   sh package-mapbox-maps.sh

Reviewer: I recommend `./package-mapbox-maps.sh` and letting the hashbang determine which shell to use.

Author: sounds good, updated.

  XCFramework for `MapboxMaps`.
- NOTE: This script **MUST** be called from this directory

### `download-dependency.sh`
- Usage:
  ```
  sh download-dependency.sh <SDK_REGISTRY_NAME> <SDK_REGISTRY_ARTIFACT> <VERSION>
  ```
- This script downloads `MapboxCoreMaps.xcframework`, `MapboxCommon.xcframework`
  and `MapboxMobileEvents.xcframework` from SDK Registry
- NOTE: You **MUST** have a valid `.netrc` token

### `build-dependency.sh`

- Usage:
  ```
  sh build-dependency.sh <NAME> <GIT_REPO_URL> <GIT_TAG> <SCHEME>
  ```
- This script clones a given repository, checks out a `git tag`, and builds the specified `scheme`.
- Uses `xcodebuild`

> - The `<NAME>` provided must match the name of the `.xcodeproj`.

Reviewer: Must also match the base name of the resulting .framework product built by $SCHEME
